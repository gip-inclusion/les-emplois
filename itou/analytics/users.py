from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
