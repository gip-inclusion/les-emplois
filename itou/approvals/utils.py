from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
