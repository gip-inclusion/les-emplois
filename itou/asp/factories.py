import datetime
import random

import factory

from itou.asp import models


_sample_europe_countries = [
    {"code": "101", "name": "DANEMARK", "group": "2"},
    {"code": "111", "name": "BULGARIE", "group": "2"},
    {"code": "135", "name": "PAYS-BAS", "group": "2"},
]

_sample_outside_europe_countries = [
    {"code": "212", "name": "AFGHANISTAN", "group": "3"},
    {"code": "436", "name": "BAHAMAS", "group": "3"},
    {"code": "324", "name": "CONGO", "group": "3"},
]

_sample_france = [
    {"code": "714", "name": "BORA-BORA", "group": "1"},
    {"code": "812", "name": "KOUMAC", "group": "1"},
    {"code": "737", "name": "PUKAPUKA", "group": "1"},
]

_sample_communes = [
    {"code": "64483", "name": "SAINT-JEAN-DE-LUZ"},
    {"code": "97108", "name": "CAPESTERRE-DE-MARIE-GALANT"},
    {"code": "97107", "name": "CAPESTERRE-BELLE-EAU"},
    {"code": "37273", "name": "VILLE-AUX-DAMES"},
    {"code": "13200", "name": "MARSEILLE"},
    {"code": "67152", "name": "GEISPOLSHEIM"},
    {"code": "85146", "name": "MONTAIGU"},
]


class CountryFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = models.Country
        django_get_or_create = ("code",)

    code, name, group = random.choice(_sample_europe_countries).values()


class CountryFranceFactory(CountryFactory):
    code, name, group = "100", "FRANCE", "1"


class CountryCollectiviteOutremerFactory(CountryFactory):
    """
    Parts or regions of France having their own INSEE country code

    Most of them are "Collectivit√©s d'Outre-Mer"
    """

    code, name, group = random.choice(_sample_france).values()


class CountryEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_europe_countries).values()


class CountryOutsideEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_outside_europe_countries).values()


class CommuneFactory(factory.django.DjangoModelFactory):
    """Factory for ASP INSEE commune:
    - if `code` or `name` are build parameters, object values will result of a lookup in `_sample_communes`
    - otherwise, fields `code` and `name` will be set from a randomly picked sample commune
    """

    class Meta:
        model = models.Commune
        django_get_or_create = (
            "code",
            "start_date",
            "end_date",
        )

    # FIXME: may cause issues in testing validity periods
    start_date = datetime.date(2000, 1, 1)
    end_date = None

    # Skipping definition of `code` and `name` fields
    # will be set in `_adjust_kwargs`

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs


# FIXME: unreliable and confusing
class MockedCommuneFactory(CommuneFactory):
    """
    A factory with a specific code for mock testing
    """

    code, name = _sample_communes[-1].values()
