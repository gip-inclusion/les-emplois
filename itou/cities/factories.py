import json
import os

from django.contrib.gis.geos import GEOSGeometry, Point
from django.template.defaultfilters import slugify

from itou.cities.models import City


CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))

CITIES_JSON_FILE = f"{CURRENT_DIR}/management/commands/data/cities.json"


def create_test_cities(selected_departments, num_per_department=None):
    """
    Not a factory strictly speaking, but it's sitting here for discoverability.

    Pick cities from `cities.json` based on the given department numbers.
        create_test_cities(['01', '02'])

    The number of cities in each department can be limited:
        create_test_cities(['54', '57'], num_per_department=2)
    """
    departments_counter = {department: 0 for department in selected_departments}

    with open(CITIES_JSON_FILE, "r") as raw_json_data:

        json_data = json.load(raw_json_data)

        for item in json_data:

            department = item.get("codeDepartement")
            if (not department) or (department not in selected_departments):
                continue

            coords = item.get("centre")
            if not coords:
                continue

            departments_counter[department] += 1
            if num_per_department and departments_counter[department] > num_per_department:
                if all(value >= num_per_department for value in departments_counter.values()):
                    break
                continue

            City.objects.create(
                slug=slugify(f"{item['nom']}-{department}"),
                department=department,
                name=item["nom"],
                post_codes=item["codesPostaux"],
                code_insee=item["code"],
                coords=GEOSGeometry(f"{coords}"),  # Feed `GEOSGeometry` with GeoJSON.
            )


def create_city_saint_andre():
    return City.objects.create(
        name="Saint-André-des-Eaux",
        slug="saint-andre-des-eaux-44",
        department="44",
        coords=Point(-2.3140436, 47.3618584),
        post_codes=["44117"],
        code_insee="44117",
    )


def create_city_guerande():
    return City.objects.create(
        name="Guérande",
        slug="guerande-44",
        department="44",
        coords=Point(-2.4747713, 47.3358576),
        # Dummy
        post_codes=["44350"],
        code_insee="44350",
    )


def create_city_vannes():
    return City.objects.create(
        name="Vannes",
        slug="vannes-56",
        department="56",
        coords=Point(-2.8186843, 47.657641),
        # Dummy
        post_codes=["56000"],
        code_insee="56000",
    )
