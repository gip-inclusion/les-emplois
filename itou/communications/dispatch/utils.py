class JobSeekerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_job_seeker


class EmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_employer


class PrescriberNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_prescriber


class PrescriberOrEmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and (self.user.is_prescriber or self.user.is_employer)
