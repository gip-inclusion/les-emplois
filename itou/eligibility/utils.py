from collections import Counter


def geiq_allowance_amount(is_authorized_prescriber, administrative_criteria) -> int:
    """Amount of granted allowance for job seeker.

    Calculated in function of:
        - author kind
        - number, annex and level of administrative criteria.
    Currently, only 3 amounts possible:
        - 0
        - 814EUR
        - 1400EUR
    """

    # Even when no criteria in the diagnosis
    if is_authorized_prescriber:
        return 1400

    # Count by annex
    annex_cnt = Counter(c.annex for c in administrative_criteria)

    # Only annex 2 administrative criteria have a level defined
    level_cnt = Counter(c.level for c in administrative_criteria if c.level)

    # At least one level 1 criterion or at least two level 2 criteria
    if level_cnt["1"] > 0 or level_cnt["2"] > 1:
        return 1400

    # Criteria in both annex ("1+2") must be counted as annex 1
    if annex_cnt["1"] + annex_cnt["1+2"] > 0:
        return 814

    return 0
