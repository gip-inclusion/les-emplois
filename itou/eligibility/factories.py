import factory
from django.utils import timezone

from itou.eligibility import models
from itou.prescribers.factories import PrescriberOrganizationWithMembershipFactory
from itou.siaes.factories import SiaeFactory
from itou.users.factories import JobSeekerFactory


class EligibilityDiagnosisFactory(factory.django.DjangoModelFactory):
    """Generate an EligibilityDiagnosis() object whose author is an authorized prescriber organization."""

    class Meta:
        model = models.EligibilityDiagnosis

    author = factory.LazyAttribute(lambda obj: obj.author_prescriber_organization.members.first())
    author_kind = models.EligibilityDiagnosis.AUTHOR_KIND_PRESCRIBER
    author_prescriber_organization = factory.SubFactory(PrescriberOrganizationWithMembershipFactory, authorized=True)
    job_seeker = factory.SubFactory(JobSeekerFactory)


class EligibilityDiagnosisMadeBySiaeFactory(factory.django.DjangoModelFactory):
    """Generate an EligibilityDiagnosis() object whose author is an SIAE."""

    class Meta:
        model = models.EligibilityDiagnosis

    author = factory.LazyAttribute(lambda obj: obj.author_siae.members.first())
    author_kind = models.EligibilityDiagnosis.AUTHOR_KIND_SIAE_STAFF
    author_siae = factory.SubFactory(SiaeFactory, with_membership=True)
    job_seeker = factory.SubFactory(JobSeekerFactory)


class ExpiredEligibilityDiagnosisFactory(EligibilityDiagnosisFactory):

    created_at = factory.LazyAttribute(
        lambda self: models.EligibilityDiagnosis.get_expiration_dt() - timezone.timedelta(days=1)
    )


class ExpiredEligibilityDiagnosisMadeBySiaeFactory(EligibilityDiagnosisMadeBySiaeFactory):

    created_at = factory.LazyAttribute(
        lambda self: models.EligibilityDiagnosis.get_expiration_dt() - timezone.timedelta(days=1)
    )


class AdministrativeCriteriaFactory(factory.django.DjangoModelFactory):
    """
    The AdministrativeCriteria table is automatically populated with a fixture at the end of
    eligibility's migration #0003.
    """

    pass
