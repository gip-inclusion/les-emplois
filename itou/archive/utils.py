import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
