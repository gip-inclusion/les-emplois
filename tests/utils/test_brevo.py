import httpx
import pytest
from django.conf import settings

from itou.utils.brevo import BrevoClient
from tests.users.factories import JobSeekerFactory


@pytest.fixture(name="brevo_client")
def brevo_client_fixture(settings):
    settings.BREVO_API_KEY = "BREVO_API_KEY"
    return BrevoClient()


@pytest.fixture(name="mock_httpx_client")
def mock_httpx_client_fixture(mocker):
    mock_client = mocker.Mock()
    mocker.patch("itou.utils.brevo.httpx.Client", return_value=mock_client)
    return mock_client


def test_brevo_client_context_manager(mock_httpx_client, caplog, brevo_client):
    with brevo_client:
        assert brevo_client.client == mock_httpx_client

    mock_httpx_client.close.assert_called_once()


def test_brevo_client_context_manager_with_exception(mock_httpx_client, caplog, brevo_client):
    with pytest.raises(Exception, match="Test exception"):
        with brevo_client:
            raise Exception("Test exception")

    mock_httpx_client.close.assert_called_once()

    error_record = next(record for record in caplog.records if record.levelname == "ERROR")
    assert "An exception occurred in BrevoClient: Test exception" in error_record.message


@pytest.mark.parametrize("status_code,content", [(202, "OK"), (400, "Bad Request"), (500, "Internal Server Error")])
def test_import_users(respx_mock, caplog, status_code, content, brevo_client):
    user = JobSeekerFactory()
    payload = {
        "listIds": [31],
        "emailBlacklist": False,
        "smsBlacklist": False,
        "updateExistingContacts": False,
        "emptyContactsAttributes": False,
        "jsonBody": [{"email": user.email}],
    }

    respx_mock.post(f"{settings.BREVO_API_URL}/contacts/import", json=payload).mock(
        return_value=httpx.Response(status_code, content=content)
    )

    brevo_client.import_users([user], 31, lambda x: {"email": x.email})

    assert respx_mock.calls.called

    if status_code != 202:
        error_record = next(record for record in caplog.records if record.levelname == "ERROR")
        assert (
            f"Brevo API: Some emails were not imported, status_code={status_code}, content={content}"
            in error_record.message
        )


def test_import_contacts_request_error(mocker, caplog, brevo_client):
    mocker.patch("itou.utils.brevo.httpx.Client.post", side_effect=httpx.RequestError("Connection timed out"))

    with pytest.raises(httpx.RequestError):
        brevo_client._import_contacts([{"email": "user1@example.com"}], 1, lambda x: x)

    error_record = next(record for record in caplog.records if record.levelname == "ERROR")
    assert error_record.message == "Brevo API: Request failed: Connection timed out"


def test_delete_contact_request_error(mocker, caplog, snapshot, brevo_client):
    email = "somebody@mail.com"
    mocker.patch("itou.utils.brevo.httpx.Client.delete", side_effect=httpx.RequestError("Connection timed out"))

    with pytest.raises(httpx.RequestError):
        # using the client directly to simulate the error, as async_delete_contact does not propagate it,
        # but catch it for managing retries
        brevo_client.delete_contact(email)

    error_record = next(record for record in caplog.records if record.levelname == "ERROR")
    assert error_record.message == snapshot(name="brevo-api-request-error-connection-timed-out")
    assert email not in caplog.text


def test_delete_contact_on_http_status_error(respx_mock, caplog, snapshot, brevo_client):
    email = "somebody@mail.com"
    respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
        return_value=httpx.Response(status_code=500)
    )

    with pytest.raises(httpx.HTTPStatusError):
        # using the client directly to simulate the error, as async_delete_contact does not propagate it
        # but catch it for managing retries
        brevo_client.delete_contact(email)

    error_record = next(record for record in caplog.records if record.levelname == "ERROR")
    assert error_record.message == snapshot(name="brevo-api-http-error-500")
    assert email not in caplog.text


def test_delete_contact_on_400(respx_mock, caplog, brevo_client):
    email = "accent@d√©monstration.fr"
    respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
        return_value=httpx.Response(
            status_code=400, json={"code": "invalid_parameter", "message": "Invalid email address"}
        )
    )

    # should not raise
    brevo_client.delete_contact(email)
    assert respx_mock.calls.called
    assert "Brevo API: email considered as invalid - no need to delete it" in caplog.text

    email = "test@example.com"
    respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
        return_value=httpx.Response(status_code=400, json={"code": "other_error", "message": "Some other error"})
    )
    with pytest.raises(httpx.HTTPStatusError):
        brevo_client.delete_contact(email)
    assert respx_mock.calls.called
