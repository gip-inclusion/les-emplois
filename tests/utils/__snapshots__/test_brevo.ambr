# serializer version: 1
# name: test_delete_contact_on_http_status_error[brevo-api-http-error-500]
  'Brevo API: Response with status_code=500'
# ---
# name: test_delete_contact_request_error[brevo-api-request-error-connection-timed-out]
  'Brevo API: Request failed: Connection timed out'
# ---
